src/ucp/tag/rndv.c

```diff
@@ -847,80 +1001,183 @@ static ucs_status_t ucp_rndv_pipeline(ucp_request_t *sreq, ucp_rndv_rtr_hdr_t *r
+        freq = ucp_request_get(worker);
+        if (freq == NULL) {
+            ucs_fatal("failed to allocate fragment receive request");
         }
 
-        mdesc = ucs_mpool_get_inline(&worker->rndv_frag_mp);
-        if (mdesc == NULL) {
-            status = UCS_ERR_NO_MEMORY;
-            goto out;
+        if (UCP_MEM_IS_HOST(sreq->send.mem_type)) {
+            /* sbuf is in host, directly do put */
+            ucp_request_send_state_reset(freq, ucp_rndv_frag_send_put_completion,
+                                         UCP_REQUEST_SEND_PROTO_RNDV_PUT);
+            freq->send.ep                         = fsreq->send.ep;
+            freq->send.buffer                     = fsreq->send.buffer + offset;
+            freq->send.datatype                   = ucp_dt_make_contig(1);
+            freq->send.mem_type                   = UCS_MEMORY_TYPE_HOST;
+            freq->send.state.dt.dt.contig.memh[0] = sreq->send.state.dt.dt.contig.memh[0];
+            freq->send.state.dt.dt.contig.md_map  = sreq->send.state.dt.dt.contig.md_map;
+            freq->send.length                     = length;
+            freq->send.uct.func                   = ucp_rndv_progress_rma_put_zcopy;
+            freq->send.rndv_put.sreq              = fsreq;
+            freq->send.rndv_put.rkey              = fsreq->send.rndv_put.rkey;
+            freq->send.rndv_put.uct_rkey          = fsreq->send.rndv_put.uct_rkey;
+            freq->send.rndv_put.remote_address    = rndv_rtr_hdr->address + offset;
+            freq->send.rndv_put.remote_request    = rndv_rtr_hdr->rreq_ptr;
+            freq->send.lane                       = fsreq->send.lane;
+            freq->send.mdesc                      = NULL;
+        } else {
+            /* perform get on memtype endpoint to stage data to host memory */
+            mem_type_ep       = worker->mem_type_ep[sreq->send.mem_type];
+            mem_type_rma_lane = ucp_ep_config(mem_type_ep)->key.rma_bw_lanes[0];
+            if (mem_type_rma_lane == UCP_NULL_LANE) {
+                return UCS_ERR_UNSUPPORTED;
+            }
+
+            mdesc = ucp_worker_mpool_get(&worker->rndv_frag_mp);
+            if (mdesc == NULL) {
+                status = UCS_ERR_NO_MEMORY;
+                goto out;
+            }
+
+            ucp_request_send_state_init(freq, ucp_dt_make_contig(1), 0);
+            ucp_request_send_state_reset(freq, ucp_rndv_frag_get_completion,
+                                         UCP_REQUEST_SEND_PROTO_RNDV_GET);
+            md_index                             = ucp_ep_md_index(mem_type_ep, mem_type_rma_lane);
+            freq->send.ep                        = mem_type_ep;
+            freq->send.buffer                    = mdesc + 1;
+            freq->send.datatype                  = ucp_dt_make_contig(1);
+            freq->send.mem_type                  = sreq->send.mem_type;
+            freq->send.state.dt.dt.contig.memh[0]= ucp_memh2uct(mdesc->memh, md_index);
```

Comment:
don't have space line between the field name and `=`

Reply:
fixed